src/CfgMgr32.Tests/CfgMgr32Facts.cs

```diff
@@ -0,0 +1,149 @@
+
+        Assert.Equal(CONFIGRET.CR_SUCCESS, CM_Register_Notification(
+            CM_NOTIFY_FILTER.Create(),
+            null,
+            callback,
+            out SafeNotificationHandle context));
+
+        context.Dispose();
+        GC.KeepAlive(callback);
+    }
+
+    // These unit tests register for notifcations, but don't process any notifications.
+    // You can test this locally by having each unit test block indefinitively before
+    // context.Dispose(). Then, trigger a device event (e.g. by unplugging a USB device),
+    // and observe the notification in the callback.
+    [Fact]
+    public unsafe void RegisterUnregisterClassGuidTest()
+    {
+        var callback = new CM_NOTIFY_CALLBACK(this.ProcessNotification);
+
+        // Register for USB events
+        Assert.Equal(CONFIGRET.CR_SUCCESS, CM_Register_Notification(
+            CM_NOTIFY_FILTER.Create(new Guid("A5DCBF10-6530-11D2-901F-00C04FB951ED")),
+            null,
+            callback,
+            out SafeNotificationHandle context));
+
+        context.Dispose();
+        GC.KeepAlive(callback);
+    }
+
+    [Fact]
+    public unsafe void RegisterUnregisterInstanceIdTest()
+    {
+        var callback = new CM_NOTIFY_CALLBACK(this.ProcessNotification);
+        var filter = CM_NOTIFY_FILTER.Create(@"ROOT\BASICDISPLAY\0000");
+
+        Assert.Equal(CONFIGRET.CR_SUCCESS, CM_Register_Notification(
+            &filter,
+            null,
+            callback,
+            out SafeNotificationHandle context));
+
+        context.Dispose();
+        GC.KeepAlive(callback);
+    }
+
+    [Fact]
+    public unsafe void RegisterUnregisterAllInterfaceClassesTest()
+    {
+        var callback = new CM_NOTIFY_CALLBACK(this.ProcessNotification);
+
+        Assert.Equal(CONFIGRET.CR_SUCCESS, CM_Register_Notification(
+            CM_NOTIFY_FILTER.AllInterfaces,
+            null,
+            callback,
+            out SafeNotificationHandle context));
+
+        context.Dispose();
+        GC.KeepAlive(callback);
+    }
+
+    [Fact]
+    public unsafe void RegisterUnregisterAllDevicesTest()
+    {
+        var callback = new CM_NOTIFY_CALLBACK(this.ProcessNotification);
+
+        Assert.Equal(CONFIGRET.CR_SUCCESS, CM_Register_Notification(
+            CM_NOTIFY_FILTER.AllDevices,
+            null,
+            callback,
+            out SafeNotificationHandle context));
+
+        context.Dispose();
+        GC.KeepAlive(callback);
+    }
+
+    [Fact]
+    public unsafe void RegisterErrorTest()
+    {
+        var callback = new CM_NOTIFY_CALLBACK(this.ProcessNotification);
+        var filter = CM_NOTIFY_FILTER.Create(@"__INVALID__");
+
+        Assert.Equal(CONFIGRET.CR_FAILURE, CM_Register_Notification(
+            &filter,
+            null,
+            callback,
+            out SafeNotificationHandle context));
+
+        Assert.True(context.IsInvalid);
+    }
+
+    private unsafe Win32ErrorCode ProcessNotification(
```

Comment:
This method isn't being executed in the tests, so the GetInstanceId and GetSymbolicLink methods aren't being tested. Is there any way to modify the test to get code coverage on these critical bits?

Reply:
Locally, the easiest way to trigger this is to have the tests block & unplug a USB device.

In CI, there isn't an easy way that I know of. You could try to mount a VHD and see if that triggers a device event, or enable/disable one of the devices which are available in the VM. (I guess there's at least a simulated mouse & keyboard available, but don't really know).

Reply:
K. I'll add a code comment hinting at this, manually test the commit pushed locally and call it good. Thanks for the tip.